### ISSUE TITLE
Bulkrax x Embargos/Leases

### ISSUE THREAD

**Issue by ShanaLMoore** [scientist-softserv/palni-palci#1010]
# Summary
[SOW](https://docs.google.com/document/d/1skNMBUHUDGRE42oNY8DvTW0ymqqLQqW-NyJ8Lz7x-_M/edit)

SoftServ will expand Bulkrax functionality by allowing the input of Embargo (and lease?) rules via CSV.  Fields will include:

- Restricted To - this text field defines the visibility level during Embargo
- Restricted Until - this date field defines the end date of the Embargo
- Open To After Restriction - this text field defines the visibility level upon expiration of the Embargo

Visibility levels will correspond with current levels in Hyku Commons:  Public, Institutional, Private.  Date inputs will follow a single standard format; functionality will not be created to convert dates or handle multiple formats.
	This work will be contributed back to Bulkrax and Hyku by SoftServ with acknowledgement to Hyku for Consortia and the University of Pittsburgh.

# Acceptance Criteria

- [ ] As an admin user I can upload Embargo rules via a CSV through Bulkrax
- [ ] CSV fields for Embargoes include:
  - [ ] Restricted To - this text field defines the visibility level during Embargo
  - [ ] Restricted Until - this date field defines the end date of the Embargo
  - [ ] Open To After Restriction - this text field defines the visibility level upon expiration of the Embargo
- [ ] This has been contributed back to Bulkrax and Hyku

# Screenshots or Video

# Testing Instructions

# Notes

###  Embargo format idea (per team meeting)

[pitt-embargo.csv](https://github.com/scientist-softserv/palni-palci/files/15001315/pitt-embargo.csv)

```csv
model,title,creator,rights_statement,bulkrax_identifier,visibility_during_embargo,embargo_release_date,visibility_after_embargo
Image,An Image,KKW,http://rightsstatements.org/vocab/InC/1.0,123456789,restricted,2024-04-16,open
```

### Lease methods/options

```ruby
work.visibility
work.visibility=
work.visibility_after_embargo
work.visibility_after_embargo=
work.visibility_after_lease
work.visibility_after_lease=
work.visibility_changed?
work.visibility_during_embargo
work.visibility_during_embargo=
work.visibility_during_lease
work.visibility_during_lease=
work.lease_expiration_date
work.lease_expiration_date=
work.lease_history
work.lease_history=
work.lease_id
work.lease_id=
work.lease_indexer_class
work.lease_visibility!
```
so header proposal: 
visibility_during_lease lease_expiration_date and visibility_after_lease

## Edge case
Validations should be considered for the following scenarios:
- [ ]

**Comment by ShanaLMoore** [comment]
Manually created work: 
<img width="883" alt="image" src="https://github.com/scientist-softserv/palni-palci/assets/10081604/ce81cc96-677e-4129-a7ac-0eaad87521b3">

**Comment by ShanaLMoore** [comment]
Current bulkrax options from the parser: 

<img width="1134" alt="image" src="https://github.com/scientist-softserv/palni-palci/assets/10081604/b0be35d0-c0e2-47ab-9700-ac414b49c95c">


[Visibility](https://github.com/samvera/bulkrax/blob/1048453a4d99a6fba4b10531c2dc3a67a51c4c50/app/parsers/bulkrax/application_parser.rb#L153-L160) via csv options: 

```ruby

    # The visibility of the record.  Acceptable values are: "open", "embaro", "lease", "authenticated", "restricted".  The default is "open"
    #
    # @return [String]
    # @see https://github.com/samvera/hydra-head/blob/main/hydra-access-controls/app/models/concerns/hydra/access_controls/access_right.rb Hydra::AccessControls::AccessRight for details on the range of values.
    # @see https://github.com/samvera/hyrax/blob/bd2bcffc33e183904be2c175367648815f25bc2b/app/services/hyrax/visibility_intention.rb Hyrax::VisibilityIntention for how we process the visibility.
    def visibility
      @visibility ||= self.parser_fields['visibility'] || 'open'
    end
```

**Comment by ShanaLMoore** [comment]
TODO: test if it just works by adding to the CSV: 
"lease_expiration_date_dtsi":"2024-04-17T00:00:00Z", "visibility_after_lease_ssim":["restricted"], "visibility_during_lease_ssim":["open"],

**Comment by ShanaLMoore** [comment]
Blocked merging it in for PITT until pals knapsacking is done.

**Comment by ShanaLMoore** [comment]
TODO: update hyku, update pals knapsack before deploying to staging

**Comment by jillpe** [comment]
Will test it once the IIIF splitting fix is merged

**Comment by ShanaLMoore** [comment]
# QA RESULTS: Pass ✅ 

Tested on: [Staging](https://demo.hyku-demo.notch8.cloud/importers/2/entries/2?locale=en)

- [ ] As an admin user I can upload Embargo rules via a CSV through Bulkrax
- [ ] CSV fields for Embargoes include:
  - [ ] Restricted To - this text field defines the visibility level during Embargo
  - [ ] Restricted Until - this date field defines the end date of the Embargo
  - [ ] Open To After Restriction - this text field defines the visibility level upon expiration of the Embargo



![Image](https://github.com/scientist-softserv/palni-palci/assets/10081604/62472170-490c-4bf4-a914-1ae9ce7d7335)



![Image](https://github.com/scientist-softserv/palni-palci/assets/10081604/83fdd3b6-716a-4cf5-b0e1-43d867354f09)



![Image](https://github.com/scientist-softserv/palni-palci/assets/10081604/e0fe00c5-7995-4580-9b69-d3a1aa59dd64)

**Comment by ShanaLMoore** [comment]
# QA RESULTS: PASS ✅ 

Tested On: [STAGING](https://demo.hyku-demo.notch8.cloud/importers/4?locale=en)

Sample: [embargo-lease.csv](https://github.com/scientist-softserv/palni-palci/files/15300412/embargo-lease.csv)




![Image](https://github.com/scientist-softserv/palni-palci/assets/10081604/d8d14e01-6dd9-4f7e-92f1-45cde1b99e47)




![Image](https://github.com/scientist-softserv/palni-palci/assets/10081604/62a41ed3-e468-4dcf-9f75-ca15c4d79449)




![Image](https://github.com/scientist-softserv/palni-palci/assets/10081604/ff896763-be8f-4923-8c6e-6059c4102e7e)

**Comment by ShanaLMoore** [comment]
sample file: 

[split-embargo-pals.zip](https://github.com/user-attachments/files/15930011/split-embargo-pals.zip)

**Comment by ShanaLMoore** [comment]
New Issue since updating submodule: https://demo.palni-palci-staging.notch8.cloud/importers/7/entries/8?locale=en



![Image](https://github.com/scientist-softserv/palni-palci/assets/10081604/d7a4160e-4593-4e65-9328-25d6c34e9e34)

**Comment by jillpe** [comment]
Blocked: There are failing specs and we need to update the submodule

**Comment by jillpe** [comment]
SoftServ QA: ✅ 
File used for import: [split-embargo-pals.zip](https://github.com/user-attachments/files/16618322/split-embargo-pals.zip)
[Importer](https://demo.palni-palci-staging.notch8.cloud/importers/25?locale=en)



https://github.com/user-attachments/assets/637d1978-e3b0-4d31-8071-5442e8cd19de

**Comment by ctgraham** [comment]
I tried submitting two imports:
https://demo.palni-palci-staging.notch8.cloud/importers/28?locale=en
and
https://demo.palni-palci-staging.notch8.cloud/importers/30?locale=en

Both listed a "Total Works:" count as expected, but neither lists any Works, and neither presents any error messages.

**Comment by ShanaLMoore** [comment]
TODO: Look into why jobs don't appear to be running in staging

**Comment by ShanaLMoore** [comment]
@ctgraham 

For your uploads, I am seeing this error in the logs.

I agree that not displaying the error is problematic, but could you fix the spreadsheets and try again?

```
[["status_message", "Failed"], ["error_class", "StandardError"], ["error_message", "failed_validation - Resource type can't be blank"]
```

**Comment by ShanaLMoore** [comment]
Handing this back to cc @laritakr  to rework the combined_files logic. STEPS TO REPRODUCE

- Locally, update submodule remote so that pals gets the latest bulkrax code (which contains bug fixes they'll need).
- Download the original file from clinton's [importer](https://demo.palni-palci-staging.notch8.cloud/importers/30?locale=en)
- Make sure required fields are there, like type and keyword. (I also updated headers: degree_name => degree and degree_level => level per pals' bulkrax field mappings).
- at this point if you upload you'll get a **FILE NOT FOUND ERROR:**
  - Bulkrax does NOT support files with spaces so I added underscores to the filename in the item column of the csv. I also renamed the file in the files directory of the zip to include underscores. 
  - rezip the file and import it
Error: 
```
[["status_message", "Failed"], ["error_class", "NoMethodError"], ["error_message", "undefined method `[]' for nil:NilClass"], ["error_backtrace", "---\n- \"/usr/local/bundle/bundler/gems/bulkrax-cd9b58226ef2/app/factories/bulkrax/valkyrie_object_factory.rb:285:in\n
```

The problem is combined_files is an empty array for this scenario.

**Comment by laritakr** [comment]
TODO: Update Bulkrax to released version and retest

**Comment by kirkkwang** [comment]
# QA PENDING

**Comment by ShanaLMoore** [comment]
# QA RESULTS: ✅ PASS

tested on [STAGING](https://demo.palni-palci-staging.notch8.cloud/importers/43?locale=en)


# Acceptance Criteria

- [ ] As an admin user I can upload Embargo rules via a CSV through Bulkrax
- [ ] CSV fields for Embargoes include:
  - [ ] Restricted To - this text field defines the visibility level during Embargo
  - [ ] Restricted Until - this date field defines the end date of the Embargo
  - [ ] Open To After Restriction - this text field defines the visibility level upon expiration of the Embargo
- [ ] This has been contributed back to Bulkrax and Hyku

### LEASE created from bulkrax

<img width="1071" alt="Image" src="https://github.com/user-attachments/assets/3a5d887f-cbd4-40d9-8371-de9ec5d7ef2c" />


### EMBARGO created from bulkrax

<img width="1070" alt="Image" src="https://github.com/user-attachments/assets/7b34e2fb-9a60-4edb-b1f9-53d296169875" />

**Comment by jillpe** [comment]
@ctgraham can you confirm if this is working as expected now?

**Comment by laritakr** [comment]
> for entries that have embargo metadata set, such as (from a JSON-formatted debug file): ` "embargo_release_date": "2025-08-27", "visibility": "embargo", "visibility_after_embargo": "open"`
> 
> What do I need to be doing instead, here?
> 

Does your csv include a column for `visibility_during_embargo`?

**Comment by alex-wreschnig** [comment]
@!*(&^!*@*(&#^!! that's a really good catch - my brain was telling me I'd included it, but no, I've accidentally refactored it out of our export after we had a change to the data format I was receiving. Let me see about getting it back in there.

Thanks! Appreciate the extra set of eyes...

**Comment by alex-wreschnig** [comment]
Pitt ULS can confirm that embargo functionality is working as expected!
